### PR TITLE
[DMS] Add address fields, allow custom addresses

### LINKED ISSUE #282: [DMS] Add address text fields to billing step

### PR DESCRIPTION
**Pull request by ML-Chen**
Here's what it looks like right now:
image
Basically you can type in the "Billing Address" box and if you click on one of the options, it'll autofill that field and the city, state, and country.
I modified LocationAutocompleteField.tsx a good amount to customize it for DMS, and ~~later I'll have to add a prop to enable those customizations for DMS so that in VMS it'll function the same as before.~~ done
Resolves #282

### CODE REVIEW

**Comment by ML-Chen** [comment]
Regarding how the form jumps up a bit when you click on the button and don't have all the fields filled out, I think that may have to do with how the form is too tall relative to the card, as when I remove the last line of form fields, it doesn't have that issue. @PC98  Do you know what code to change to make the card taller?

**Comment by ML-Chen** [comment]
I was trying to use TypeScript generics so that LocationAutocompleteInput can use `PlaceType | string` if you use it allowing free-form text, and `PlaceType` if you don't. Unfortunately, it doesn't seem that TypeScript can support using generics like that. You can't use booleans like `T extends true` to control the types, it seems; see https://stackoverflow.com/questions/51432121/typescript-conditional-types-and-using-a-boolean-parameter-to-control-the-retur. I tried doing `<T extends (PlaceType | string) | PlaceType = PlaceType`, but TypeScript doesn't let you assign values to generics because it ["could be instantiated with a different type of constraint"](https://stackoverflow.com/questions/56505560/could-be-instantiated-with-a-different-subtype-of-constraint-object). So I had to do some `as PlaceType` type assertions.

File: components/donation/form/DonationPageForm.tsx [hunk 1]
```diff
@@ -142,22 +142,21 @@ const DonationPageForm: React.FC<Props> = ({
...
             billingStep.email,
             billingStep.zipcode,
-            city,
-            billingStep.address.structured_formatting.main_text,
-            state,
-            country.slice(0, -1)
+            billingStep.city,
+            billingStep.addressLine,
+            billingStep.state,
+            billingStep.country.slice(0, -1)
```
Comment (PC98) [inline components/donation/form/DonationPageForm.tsx:170]: Why the `.slice(0, -1)`?

File: components/donation/form/DonationPageForm.tsx [hunk 2]
```diff
@@ -142,22 +142,21 @@ const DonationPageForm: React.FC<Props> = ({
...
-            country
-          ] = billingStep.address.structured_formatting.secondary_text.split(
-            ", "
-          );
+        if (
+          isPaymentStep &&
+          billingStep.addressLine &&
+          billingStep.city &&
+          billingStep.state &&
+          billingStep.country
```
Comment (PC98) [inline components/donation/form/DonationPageForm.tsx:161]: Do we need these conditions on `billingStep`, given that `DonationPageFormBillingStep.tsx` marks all these inputs as required?

File: server/migrations/20201018194031-migrate-events-address-format.ts [hunk 3]
```diff
@@ -28,7 +28,7 @@ export const up: MigrationFunction = async (db: Db) => {
           $set: {
             "address.text": {
               main: faker.random.arrayElement(MAIN_TEXT_OPTIONS),
-              // @ts-ignore: address.text is of type string
+              // @ts-ignore: addressLine.text is of type string
```
Comment (PC98) [inline server/migrations/20201018194031-migrate-events-address-format.ts:31]: I don't think this change is related to your PR? Also ideally we should not modify already committed migration scripts

File: components/donation/form/reducer.ts [hunk 4]
```diff
@@ -13,7 +13,11 @@ export interface BillingStepProps {
   firstName: string;
   lastName: string;
   email: string;
-  address: PlaceType | null;
+  addressField: PlaceType | string | null;
```
Comment (PC98) [inline components/donation/form/reducer.ts:16]: If we are okay with `freeSolo`, why do we need to store `addressField` in our redux store? Doesn't `addressLine` and other address fields suffice?

File: components/LocationAutocompleteInput.tsx [hunk 5]
```diff
@@ -37,14 +37,19 @@ export type PlaceType = google.maps.places.AutocompletePrediction;
 
 interface Props {
   locationType: string;
-  parentCallback: (value: PlaceType) => void;
-  filterOptions?: (options: PlaceType[]) => PlaceType[];
+  parentCallback: (value: PlaceType | string) => void;
```
Comment (PC98) [inline components/LocationAutocompleteInput.tsx:40]: Instead of typing `PlaceType | string` over and over again, make a new type `type LocationValue = PlaceType | string;` or whatever you wanna call it and use that.

File: components/LocationAutocompleteInput.tsx [hunk 6]
```diff
@@ -190,34 +202,36 @@ const LocationAutocompleteInput: React.FC<Props> = ({
                 <SearchIcon fontSize="inherit" />
               </InputAdornment>
             ),
-            classes: { root: textStyle }
+            ...(small && { classes: { root: textStyle } })
```
Comment (PC98) [inline components/LocationAutocompleteInput.tsx:205]: Is DMS not supposed to use caption typography style for their autocomplete? If not, the easiest thing to do might be to take text variant as a style:
```javascript
// Outside the component:
interface StyleProps {
  textVariant: keyof typeof typographyStyles;
}

const useStyles = makeStyles({
  // @ts-ignore: Pretty sure bug in mui types
  textStyle: (props: StyleProps) => typographyStyles[props.textVariant],
  // rest
});

interface Props extends StyleProps {
  // same expect remove "small"
}
// Inside the component:
const { textStyle, highlightedText, inputAdornmentRoot } = useStyles({
  textVariant
});
```

Then you can remove changes made on this line.

File: components/LocationAutocompleteInput.tsx [hunk 7]
```diff
@@ -190,34 +202,36 @@ const LocationAutocompleteInput: React.FC<Props> = ({
...
-        return (
-          <Grid container alignItems="center">
-            {parts.map(({ text, highlight }, index) => (
-              <CoreTypography
-                variant="caption"
-                key={index}
-                className={clsx(highlight && highlightedText)}
-              >
-                {text.replace(/ /g, "\u00a0")}
+        if (typeof option !== "string") {
```
Comment (PC98) [inline components/LocationAutocompleteInput.tsx:210]: I see we are comparing against `"string"` type in a lot of places. Let's make a type guard outside of this component:
```javascript
const isAutocompletePrediction = (
  value: LocationValue | null
): value is PlaceType => value != null && typeof value !== "string";
```
Then simply call it everywhere you want to check `value`'s type: `isAutocompletePrediction(value)`

File: components/LocationAutocompleteInput.tsx [hunk 8]
```diff
@@ -37,14 +37,19 @@ export type PlaceType = google.maps.places.AutocompletePrediction;
...
+    options: Array<PlaceType | string>
+  ) => Array<PlaceType | string>;
   label?: string;
   fullWidth?: boolean;
-  defaultValue?: PlaceType | null;
+  defaultValue?: PlaceType | string | null;
   required?: boolean;
   placeholder?: string;
   clearInputOnClose?: boolean;
+  freeInput?: boolean;
```
Comment (PC98) [inline components/LocationAutocompleteInput.tsx:50]: nit: use same name as prop, `freeSolo`

File: components/LocationAutocompleteInput.tsx [hunk 9]
```diff
@@ -56,12 +61,16 @@ const LocationAutocompleteInput: React.FC<Props> = ({
...
-  clearInputOnClose = false
+  clearInputOnClose = false,
+  freeInput = false,
+  outlined = true,
+  small = true
 }) => {
   const { textStyle, highlightedText, inputAdornmentRoot } = useStyles();
-  const [value, setValue] = useState<PlaceType | null>(defaultValue);
+  // may be string if freeInput is true and user selects text not in autocomplete suggestions
+  const [value, setValue] = useState<PlaceType | string | null>(defaultValue);
```
Comment (PC98) [inline components/LocationAutocompleteInput.tsx:71]: just noticed this but we don't need the type specification `<PlaceType | string | null>` here.

File: components/LocationAutocompleteInput.tsx [hunk 10]
```diff
@@ -56,12 +61,16 @@ const LocationAutocompleteInput: React.FC<Props> = ({
...
   const [inputValue, setInputValue] = useState("");
-  const [options, setOptions] = useState<PlaceType[]>([]);
+  const [options, setOptions] = useState<Array<string | PlaceType>>([]);
```
Comment (PC98) [inline components/LocationAutocompleteInput.tsx:73]: We should keep `options` as `PlaceType[]`, if we expect non-`freeSolo` to be used with autocomplete predictions. You might get a few TS errors, but they can be easily fixed using the type guard method suggested above. As an example:
```javascript
// in the useEffect
if (inputValue === "") {
  setOptions(isAutocompletePrediction(value) ? [value] : []);
  return undefined;
}
```

File: components/LocationAutocompleteInput.tsx [hunk 11]
```diff
@@ -123,8 +132,7 @@ const LocationAutocompleteInput: React.FC<Props> = ({
       },
       (results?: PlaceType[]) => {
         if (active) {
-          let newOptions = [] as PlaceType[];
-
+          let newOptions: Array<PlaceType | string> = [];
```
Comment (PC98) [inline components/LocationAutocompleteInput.tsx:135]: Undo this change and use `isAutocompletePrediction` on line 136.

File: components/LocationAutocompleteInput.tsx [hunk 12]
```diff
@@ -146,7 +154,9 @@ const LocationAutocompleteInput: React.FC<Props> = ({
   return (
     <Autocomplete
       getOptionLabel={option =>
-        typeof option === "string" ? option : option.description
+        typeof option === "string"
```
Comment (PC98) [inline components/LocationAutocompleteInput.tsx:157]: We will be able to use `isAutocompletePrediction` here as well.

File: components/donation/form/DonationPageFormBillingStep.tsx [hunk 13]
```diff
@@ -126,17 +128,53 @@ const DonationPageFormBillingStep: React.FC<BillingStepProps> = ({
           locationType="address"
           fullWidth
           required
-          defaultValue={address}
+          defaultValue={addressLine}
           label="Billing Address"
+          freeInput={true}
```
Comment (PC98) [inline components/donation/form/DonationPageFormBillingStep.tsx:133]: nit: no need for `={true}` for `true` boolean props

File: components/donation/form/reducer.ts [hunk 14]
```diff
@@ -13,7 +13,11 @@ export interface BillingStepProps {
...
+  addressLine: string | null;
```
Comment (PC98) [inline components/donation/form/reducer.ts:17]: Why not just `string`?

**Review by PC98 — COMMENTED** [review]
If you wish to re-use the `isAutocompletePrediction` type guard and `LocationValue` type I have suggested in `reducer.ts` and `EventsPageLocationFilter.tsx` feel free to put them in some `utils/autocomplete.ts` file.

File: components/donation/form/DonationPageForm.tsx (hunk 1 above)
Comment (ML-Chen) [reply components/donation/form/DonationPageForm.tsx:170]: I have no idea. That was there before.

File: server/migrations/20201018194031-migrate-events-address-format.ts (hunk 3 above)
Comment (ML-Chen) [reply server/migrations/20201018194031-migrate-events-address-format.ts:31]: Yeah that was an accident, I'll revert it.

File: components/donation/form/DonationPageForm.tsx (hunk 1 above)
Comment (KoolCards) [reply components/donation/form/DonationPageForm.tsx:170]: Stripe needs a two letter country code (US instead of USA) which is why the slice is there. This won't work for other companies, but isn't an issue rn.

Comment (PC98) [reply components/donation/form/DonationPageForm.tsx:170]: This might be problematic now though since I can enter any country in the country text box.

We should add a TODO/comment explaining this behavior and open a task to get the country code from Google's Place Details API.

**Comment by PC98** [comment]
Closing this since changes went in as part of #371
